File: ts/tests/mirror.test.ts
import { describe, it, expect, afterEach } from 'vitest';
import { freshStore } from './helpers.js';
import { GitStore } from '../src/index.js';
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as os from 'node:os';

// Helper to create a temp directory for remotes/bundles
function makeTmpDir(): string {
  return fs.mkdtempSync(path.join(os.tmpdir(), 'vost-mirror-'));
}

const cleanups: string[] = [];

afterEach(() => {
  for (const d of cleanups) {
    fs.rmSync(d, { recursive: true, force: true });
  }
  cleanups.length = 0;
});

// ---------------------------------------------------------------------------
// backup / restore via local path
// ---------------------------------------------------------------------------

describe('Mirror: local path backup/restore', () => {
  it('backup to local bare repo', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');

    const remoteDir = makeTmpDir();
    cleanups.push(remoteDir);
    const remoteUrl = path.join(remoteDir, 'remote.git');

    const diff = await store.backup(remoteUrl);
    expect(diff.add.length).toBeGreaterThan(0);

    // Verify remote has the refs
    const remote = await GitStore.open(remoteUrl, { fs });
    expect(await remote.branches.has('main')).toBe(true);
    expect(await (await remote.branches.get('main')).readText('a.txt')).toBe(
      'hello',
    );
  });

  it('restore from local bare repo', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');

    const remoteDir = makeTmpDir();
    cleanups.push(remoteDir);
    const remoteUrl = path.join(remoteDir, 'remote.git');
    await store.backup(remoteUrl);

    // Create empty store and restore
    const { store: store2, tmpDir: td2 } = await freshStore({ branch: null });
    cleanups.push(td2);

    const diff = await store2.restore(remoteUrl);
    expect(diff.add.length).toBeGreaterThan(0);
    expect(await store2.branches.has('main')).toBe(true);
    expect(
      await (await store2.branches.get('main')).readText('a.txt'),
    ).toBe('hello');
  });

  it('restore is additive', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');

    const remoteDir = makeTmpDir();
    cleanups.push(remoteDir);
    const remoteUrl = path.join(remoteDir, 'remote.git');
    await store.backup(remoteUrl);

    // Create a local-only branch
    await store.branches.set('local-only', snap);
    expect(await store.branches.has('local-only')).toBe(true);

    // Restore — local-only should survive
    const diff = await store.restore(remoteUrl);
    expect(diff.delete.length).toBe(0);
    expect(await store.branches.has('local-only')).toBe(true);
  });

  it('dry-run backup makes no changes', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');

    const remoteDir = makeTmpDir();
    cleanups.push(remoteDir);
    const remoteUrl = path.join(remoteDir, 'remote.git');
    await store.backup(remoteUrl);

    // Write more data
    snap = await store.branches.get('main');
    snap = await snap.writeText('b.txt', 'world');

    const diff = await store.backup(remoteUrl, { dryRun: true });
    expect(diff.update.length + diff.add.length).toBeGreaterThan(0);

    // Remote should still not have b.txt
    const remote = await GitStore.open(remoteUrl, { fs });
    expect(await (await remote.branches.get('main')).exists('b.txt')).toBe(
      false,
    );
  });

  it('backup deletes stale remote refs', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');
    await store.branches.set('extra', snap);

    const remoteDir = makeTmpDir();
    cleanups.push(remoteDir);
    const remoteUrl = path.join(remoteDir, 'remote.git');
    await store.backup(remoteUrl);

    {
      const remote = await GitStore.open(remoteUrl, { fs });
      expect(await remote.branches.has('extra')).toBe(true);
    }

    // Delete extra locally
    await store.branches.delete('extra');

    // Backup again — should delete from remote
    const diff = await store.backup(remoteUrl);
    expect(diff.delete.some((r) => r.ref.includes('extra'))).toBe(true);

    const remote = await GitStore.open(remoteUrl, { fs });
    expect(await remote.branches.has('extra')).toBe(false);
  });

  it('round-trip backup then restore', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'aaa');
    snap = await snap.writeText('b.txt', 'bbb');
    await store.branches.set('feature', snap);
    let feat = await store.branches.get('feature');
    feat = await feat.writeText('c.txt', 'ccc');

    const remoteDir = makeTmpDir();
    cleanups.push(remoteDir);
    const remoteUrl = path.join(remoteDir, 'remote.git');
    await store.backup(remoteUrl);

    const { store: store2, tmpDir: td2 } = await freshStore({ branch: null });
    cleanups.push(td2);
    await store2.restore(remoteUrl);

    expect(
      await (await store2.branches.get('main')).readText('a.txt'),
    ).toBe('aaa');
    expect(
      await (await store2.branches.get('main')).readText('b.txt'),
    ).toBe('bbb');
    expect(await store2.branches.has('feature')).toBe(true);
    expect(
      await (await store2.branches.get('feature')).readText('c.txt'),
    ).toBe('ccc');
  });

  it('backup when already in sync', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');

    const remoteDir = makeTmpDir();
    cleanups.push(remoteDir);
    const remoteUrl = path.join(remoteDir, 'remote.git');
    await store.backup(remoteUrl);

    const diff = await store.backup(remoteUrl);
    expect(diff.add.length + diff.update.length + diff.delete.length).toBe(0);
  });

  it('backup with tags', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');
    await store.tags.set('v1.0', snap);

    const remoteDir = makeTmpDir();
    cleanups.push(remoteDir);
    const remoteUrl = path.join(remoteDir, 'remote.git');
    await store.backup(remoteUrl);

    const remote = await GitStore.open(remoteUrl, { fs });
    expect(await remote.tags.has('v1.0')).toBe(true);
    expect(await (await remote.tags.get('v1.0')).readText('a.txt')).toBe(
      'hello',
    );
  });
});

// ---------------------------------------------------------------------------
// bundle
// ---------------------------------------------------------------------------

describe('Mirror: bundle', () => {
  it('backup to bundle', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');
    await store.tags.set('v1.0', snap);

    const bundlePath = path.join(tmpDir, 'backup.bundle');
    const diff = await store.backup(bundlePath);

    expect(diff.add.length).toBeGreaterThan(0);
    expect(fs.existsSync(bundlePath)).toBe(true);
  });

  it('restore from bundle', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');
    await store.tags.set('v1.0', snap);

    const bundlePath = path.join(tmpDir, 'backup.bundle');
    await store.backup(bundlePath);

    const { store: store2, tmpDir: td2 } = await freshStore({ branch: null });
    cleanups.push(td2);

    const diff = await store2.restore(bundlePath);
    expect(diff.add.length).toBeGreaterThan(0);
    expect(await store2.branches.has('main')).toBe(true);
    expect(
      await (await store2.branches.get('main')).readText('a.txt'),
    ).toBe('hello');
    expect(await store2.tags.has('v1.0')).toBe(true);
  });

  it('bundle dry run', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');

    const bundlePath = path.join(tmpDir, 'backup.bundle');
    const diff = await store.backup(bundlePath, { dryRun: true });

    expect(diff.add.length).toBeGreaterThan(0);
    expect(fs.existsSync(bundlePath)).toBe(false);
  });

  it('bundle round trip', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'aaa');
    snap = await snap.writeText('b.txt', 'bbb');
    await store.tags.set('v1.0', snap);

    const bundlePath = path.join(tmpDir, 'roundtrip.bundle');
    await store.backup(bundlePath);

    const { store: store2, tmpDir: td2 } = await freshStore({ branch: null });
    cleanups.push(td2);
    await store2.restore(bundlePath);

    expect(
      await (await store2.branches.get('main')).readText('a.txt'),
    ).toBe('aaa');
    expect(
      await (await store2.branches.get('main')).readText('b.txt'),
    ).toBe('bbb');
    expect(await store2.tags.has('v1.0')).toBe(true);
  });
});

// ---------------------------------------------------------------------------
// refs filtering
// ---------------------------------------------------------------------------

describe('Mirror: refs filtering', () => {
  it('backup with refs filter', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');
    await store.tags.set('v1.0', snap);

    const remoteDir = makeTmpDir();
    cleanups.push(remoteDir);
    const remoteUrl = path.join(remoteDir, 'remote.git');
    await store.backup(remoteUrl, { refs: ['main'] });

    const remote = await GitStore.open(remoteUrl, { fs });
    expect(await remote.branches.has('main')).toBe(true);
    expect(await remote.tags.has('v1.0')).toBe(false);
  });

  it('restore with refs filter', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');
    await store.tags.set('v1.0', snap);

    const remoteDir = makeTmpDir();
    cleanups.push(remoteDir);
    const remoteUrl = path.join(remoteDir, 'remote.git');
    await store.backup(remoteUrl);

    const { store: store2, tmpDir: td2 } = await freshStore({ branch: null });
    cleanups.push(td2);
    await store2.restore(remoteUrl, { refs: ['v1.0'] });

    expect(await store2.tags.has('v1.0')).toBe(true);
  });

  it('backup bundle with refs', async () => {
    const { store, tmpDir } = await freshStore();
    cleanups.push(tmpDir);
    let snap = await store.branches.get('main');
    snap = await snap.writeText('a.txt', 'hello');
    await store.tags.set('v1.0', snap);

    const bundlePath = path.join(tmpDir, 'main-only.bundle');
    await store.backup(bundlePath, { refs: ['main'] });

    const { store: store2, tmpDir: td2 } = await freshStore({ branch: null });
    cleanups.push(td2);
    await store2.restore(bundlePath);

    expect(await store2.branches.has('main')).toBe(true);
    expect(await store2.tags.has('v1.0')).toBe(false);
  });
});

File: kotlin/src/test/kotlin/vost/MirrorTest.kt
package vost

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.io.File
import java.nio.file.Path
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class MirrorTest {

    private fun createRemoteDir(tempDir: Path): String {
        val remoteDir = tempDir.resolve("remote.git").toFile()
        return remoteDir.absolutePath
    }

    @Test
    fun `backup to local bare repo`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())

            val remoteUrl = createRemoteDir(tempDir)
            val diff = it.backup(remoteUrl)

            assertFalse(diff.inSync)
            assertTrue(diff.add.isNotEmpty())

            // Verify remote has the refs
            val remote = GitStore.open(remoteUrl, create = false)
            remote.use { r ->
                val remoteBranches = r.branches.list()
                assertTrue("main" in remoteBranches)
                assertEquals("hello", r.branches["main"].readText("a.txt"))
            }
        }
    }

    @Test
    fun `restore from local bare repo`(@TempDir tempDir: Path) {
        // Create source store with data
        val store1 = createStore()
        store1.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())

            val remoteUrl = createRemoteDir(tempDir)
            it.backup(remoteUrl)

            // Create a new empty store and restore into it
            val store2Dir = tempDir.resolve("restored.git").toFile()
            val store2 = GitStore.open(store2Dir.absolutePath, branch = null)
            store2.use { s2 ->
                val diff = s2.restore(remoteUrl)
                assertFalse(diff.inSync)
                assertTrue(diff.add.isNotEmpty())

                val branches = s2.branches.list()
                assertTrue("main" in branches)
                assertEquals("hello", s2.branches["main"].readText("a.txt"))
            }
        }
    }

    @Test
    fun `dry-run backup makes no changes`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())

            val remoteUrl = createRemoteDir(tempDir)
            // First do a real backup so remote exists
            it.backup(remoteUrl)

            // Write more data
            fs = it.branches["main"]
            fs = fs.write("b.txt", "world".toByteArray())

            // Dry-run should report changes but not push
            val diff = it.backup(remoteUrl, dryRun = true)
            assertTrue(diff.update.isNotEmpty() || diff.add.isNotEmpty())

            // Remote should still only have the old data
            val remote = GitStore.open(remoteUrl, create = false)
            remote.use { r ->
                assertFalse(r.branches["main"].exists("b.txt"))
            }
        }
    }

    @Test
    fun `dry-run restore makes no changes`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())

            val remoteUrl = createRemoteDir(tempDir)
            it.backup(remoteUrl)

            // Create empty store
            val store2Dir = tempDir.resolve("restored.git").toFile()
            val store2 = GitStore.open(store2Dir.absolutePath, branch = null)
            store2.use { s2 ->
                val diff = s2.restore(remoteUrl, dryRun = true)
                assertFalse(diff.inSync)

                // Store2 should still be empty
                assertTrue(s2.branches.list().isEmpty())
            }
        }
    }

    @Test
    fun `backup deletes stale remote refs`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())

            // Create a second branch
            it.branches["extra"] = fs

            val remoteUrl = createRemoteDir(tempDir)
            it.backup(remoteUrl)

            // Verify remote has both branches
            val remote1 = GitStore.open(remoteUrl, create = false)
            remote1.use { r ->
                assertTrue("extra" in r.branches.list())
            }

            // Delete the extra branch locally
            it.branches.delete("extra")

            // Backup again — should delete the remote extra branch
            val diff = it.backup(remoteUrl)
            assertTrue(diff.delete.any { it.refName.contains("extra") })

            // Verify remote no longer has the extra branch
            val remote2 = GitStore.open(remoteUrl, create = false)
            remote2.use { r ->
                assertFalse("extra" in r.branches.list())
            }
        }
    }

    @Test
    fun `restore is additive`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())

            val remoteUrl = createRemoteDir(tempDir)
            it.backup(remoteUrl)

            // Create a local-only branch
            it.branches["local-only"] = fs
            assertTrue("local-only" in it.branches.list())

            // Restore from remote — local-only branch should survive (additive)
            val diff = it.restore(remoteUrl)
            assertTrue(diff.delete.isEmpty())
            assertTrue("local-only" in it.branches.list())
        }
    }

    @Test
    fun `round-trip backup then restore`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "aaa".toByteArray())
            fs = fs.write("b.txt", "bbb".toByteArray())

            it.branches["feature"] = fs
            var feat = it.branches["feature"]
            feat = feat.write("c.txt", "ccc".toByteArray())

            val remoteUrl = createRemoteDir(tempDir)
            it.backup(remoteUrl)

            // Create new store and restore
            val store2Dir = tempDir.resolve("restored.git").toFile()
            val store2 = GitStore.open(store2Dir.absolutePath, branch = null)
            store2.use { s2 ->
                s2.restore(remoteUrl)

                assertEquals("aaa", s2.branches["main"].readText("a.txt"))
                assertEquals("bbb", s2.branches["main"].readText("b.txt"))
                assertTrue("feature" in s2.branches.list())
                assertEquals("ccc", s2.branches["feature"].readText("c.txt"))
            }
        }
    }

    @Test
    fun `backup when already in sync`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())

            val remoteUrl = createRemoteDir(tempDir)
            it.backup(remoteUrl)

            // Second backup should be in sync
            val diff = it.backup(remoteUrl)
            assertTrue(diff.inSync)
            assertEquals(0, diff.total)
        }
    }

    @Test
    fun `backup with tags`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())
            it.tags["v1.0"] = fs

            val remoteUrl = createRemoteDir(tempDir)
            it.backup(remoteUrl)

            val remote = GitStore.open(remoteUrl, create = false)
            remote.use { r ->
                assertTrue("v1.0" in r.tags.list())
                assertEquals("hello", r.tags["v1.0"].readText("a.txt"))
            }
        }
    }

    // ── Bundle tests ────────────────────────────────────────────────────

    @Test
    fun `backup to bundle`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())
            it.tags["v1.0"] = fs

            val bundle = tempDir.resolve("backup.bundle").toFile().absolutePath
            val diff = it.backup(bundle)

            assertFalse(diff.inSync)
            assertTrue(diff.add.isNotEmpty())
            assertTrue(File(bundle).exists())
        }
    }

    @Test
    fun `restore from bundle`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())
            it.tags["v1.0"] = fs

            val bundle = tempDir.resolve("backup.bundle").toFile().absolutePath
            it.backup(bundle)

            val store2Dir = tempDir.resolve("restored.git").toFile()
            val store2 = GitStore.open(store2Dir.absolutePath, branch = null)
            store2.use { s2 ->
                val diff = s2.restore(bundle)
                assertFalse(diff.inSync)

                assertTrue("main" in s2.branches.list())
                assertEquals("hello", s2.branches["main"].readText("a.txt"))
                assertTrue("v1.0" in s2.tags.list())
            }
        }
    }

    @Test
    fun `bundle dry run`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())

            val bundle = tempDir.resolve("backup.bundle").toFile().absolutePath
            val diff = it.backup(bundle, dryRun = true)

            assertFalse(diff.inSync)
            assertFalse(File(bundle).exists())
        }
    }

    @Test
    fun `bundle round trip`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "aaa".toByteArray())
            fs = fs.write("b.txt", "bbb".toByteArray())
            it.tags["v1.0"] = fs

            val bundle = tempDir.resolve("roundtrip.bundle").toFile().absolutePath
            it.backup(bundle)

            val store2Dir = tempDir.resolve("restored.git").toFile()
            val store2 = GitStore.open(store2Dir.absolutePath, branch = null)
            store2.use { s2 ->
                s2.restore(bundle)
                assertEquals("aaa", s2.branches["main"].readText("a.txt"))
                assertEquals("bbb", s2.branches["main"].readText("b.txt"))
                assertTrue("v1.0" in s2.tags.list())
            }
        }
    }

    // ── Refs filtering tests ────────────────────────────────────────────

    @Test
    fun `backup with refs filter`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())
            it.tags["v1.0"] = fs

            val remoteUrl = createRemoteDir(tempDir)
            it.backup(remoteUrl, refs = listOf("main"))

            val remote = GitStore.open(remoteUrl, create = false)
            remote.use { r ->
                assertTrue("main" in r.branches.list())
                assertFalse("v1.0" in r.tags.list())
            }
        }
    }

    @Test
    fun `restore with refs filter`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())
            it.tags["v1.0"] = fs

            val remoteUrl = createRemoteDir(tempDir)
            it.backup(remoteUrl)

            val store2Dir = tempDir.resolve("restored.git").toFile()
            val store2 = GitStore.open(store2Dir.absolutePath, branch = null)
            store2.use { s2 ->
                s2.restore(remoteUrl, refs = listOf("v1.0"))
                assertTrue("v1.0" in s2.tags.list())
            }
        }
    }

    @Test
    fun `backup bundle with refs`(@TempDir tempDir: Path) {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "hello".toByteArray())
            it.tags["v1.0"] = fs

            val bundle = tempDir.resolve("main-only.bundle").toFile().absolutePath
            it.backup(bundle, refs = listOf("main"))

            val store2Dir = tempDir.resolve("restored.git").toFile()
            val store2 = GitStore.open(store2Dir.absolutePath, branch = null)
            store2.use { s2 ->
                s2.restore(bundle)
                assertTrue("main" in s2.branches.list())
                assertFalse("v1.0" in s2.tags.list())
            }
        }
    }
}
